scripts/md-gen/controller-config/main.go

```diff
@@ -0,0 +1,472 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/schema"
+	"gopkg.in/yaml.v3"
+
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/testing"
+)
+
+// bidirectional mapping between key and constant name
+// e.g. "agent-ratelimit-max" <--> "AgentRateLimitMax"
+// These are filled by iterating through the AST of config.go
+var (
+	keyForConstantName = map[string]string{}
+	constantNameForKey = map[string]string{}
+)
+
+// Generate Markdown documentation based on the contents of the
+// github.com/juju/juju/controller package.
+func main() {
+	outputDir := os.Getenv("DOCS_DIR")              // directory to write output to
+	packagePath := os.Getenv("CONTROLLER_PKG_PATH") // path to local github.com/juju/juju/controller package
```

Comment:
I reckon it'd be slightly nicer if these had sensible defaults: `./docs` and `./controller`. Also, would it be better to make them command line options (with the `flag` package) rather than env vars? They're a bit more self-describing that way, and you get automatic support for defaults.

Reply:
Hmm, not sure. We don't know *where* the user is going to run this script, so setting defaults seems like a good way to shoot ourselves in the foot. I'd rather require these variables to be set, and fail otherwise.

Personally I prefer the env variables to flags, so I think it's a matter of taste. I think it's simpler.

Just a thought - maybe the path provided should be to the root of the Juju source tree, not just the controller package - then we can `filepath.Join(..., "controller")` as required.

Reply:
That's fine with me, no need to debate at length -- sounds like you've thought about it.